### PR TITLE
BITAU-104 Implement BridgeService

### PR DESCRIPTION
**Pull request by ahaisting-livefront**
## 🎟️ Tracking

https://livefront.atlassian.net/browse/BITAU-104

## 📔 Objective

The goal if this PR is to setup the core syncing logic within the `BridgeServiceProcessor`, and really, namely, the `syncAccounts` call. 

## 📸 Screenshots

N/A

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt [hunk 1]
```diff
@@ -37,42 +69,160 @@ class BridgeServiceProcessorImpl(
...
+                    val decryptedUserKey = authDiskSource.getAuthenticatorSyncUnlockKey(userId)
+                        ?: return@mapNotNull null
+
+                    // Wait for any unlocking actions to finish:
+                    vaultRepository.vaultUnlockDataStateFlow.first {
+                        it.statusFor(userId) != VaultUnlockData.Status.UNLOCKING
+                    }
+
+                    // Unlock vault if necessary:
+                    // TODO: Check for never-lock functionality ???:
```
Comment (ahaisting-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt:122]: @david-livefront I'm gonna ask you a question about this later 👀 🤔

File: app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt [hunk 2]
```diff
@@ -2,22 +2,54 @@ package com.x8bit.bitwarden.data.platform.processor
...
  */
+@Suppress("LongParameterList")
 class BridgeServiceProcessorImpl(
+    private val authDiskSource: AuthDiskSource,
+    private val authRepository: AuthRepository,
     private val featureFlagManager: FeatureFlagManager,
+    private val settingsDiskSource: SettingsDiskSource,
+    private val vaultDiskSource: VaultDiskSource,
+    private val vaultRepository: VaultRepository,
+    private val vaultSdkSource: VaultSdkSource,
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt:46]: Processors shouldn't have direct access to these "sources", can we abstract this behind a `BridgeRepository` or something like that?

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+     * Users who do not have authenticator sync enabled or otherwise cannot have their ciphers
+     * accessed will be omitted from the list.
+     */
+    suspend fun getSharedAccounts(): SharedAccountData
+
+    /**
+     * The currently persisted authenticator sync symmetric key. This key is used for
+     * encrypting IPC traffic.
+     */
+    val authenticatorSyncSymmetricKey: ByteArray?
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt:26]: Vals should be above the functions

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepositoryImpl.kt [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+            SharedAccountData.Account(
+                userId = account.userId,
+                name = account.name,
+                email = account.email,
+                environmentLabel = account.environment.label,
+                lastSyncTime = lastSyncTime,
+                totpUris = totpUris,
+            )
+        }.let {
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepositoryImpl.kt:110]: This should be chained:
```kotlin
return allAccounts
    .mapNotNull { account ->
        ...
    }
    .let { SharedAccountData(it) }
```

File: app/src/test/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorTest.kt [hunk 5]
```diff
@@ -1,52 +1,212 @@
...
+        bridgeServiceProcessor = BridgeServiceProcessorImpl(
+            bridgeRepository = bridgeRepository,
+            featureFlagManager = featureFlagManager,
+            dispatcherManager = FakeDispatcherManager(),
+        )
     }
 
     @Test
     fun `when AuthenticatorSync feature flag is off, should return null binder`() {
+        mockkStatic(::isBuildVersionBelow)
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorTest.kt:52]: This should be unmocked in the tearDown

File: app/src/test/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorTest.kt [hunk 6]
```diff
@@ -1,52 +1,212 @@
...
+
+        @Test
+        fun `syncAccounts should encrypt result from BridgeRepository`() {
+            val sharedAccountData = mockk<SharedAccountData>()
+            val expected = mockk<EncryptedSharedAccountData>()
+            every {
+                bridgeRepository.authenticatorSyncSymmetricKey
+            } returns SYMMETRIC_KEY.symmetricEncryptionKey.byteArray
+            coEvery { bridgeRepository.getSharedAccounts() } returns sharedAccountData
+            mockkStatic(SharedAccountData::encrypt)
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorTest.kt:178]: This should also be unmocked in the tearDown

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepositoryTest.kt [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+            coVerify { vaultSdkSource.decryptCipher(USER_2_ID, USER_2_ENCRYPTED_SDK_TOTP_CIPHER) }
+        }
+
+    @Test
+    fun `syncAccounts when userStateFlow is null should return an empty list`() = runTest {
+        every { authRepository.userStateFlow } returns MutableStateFlow(null)
+
+        val sharedData = bridgeRepository.getSharedAccounts()
+
+        assert(sharedData.accounts.isEmpty())
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepositoryTest.kt:159]: Can we make this `assertTrue`

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepositoryTest.kt [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+            )
+
+            deferred.await()
+
+            verify { authRepository.userStateFlow }
+            verify { vaultDiskSource.getCiphers(USER_1_ID) }
+            verify { vaultDiskSource.getCiphers(USER_2_ID) }
+            verify { vaultRepository.isVaultUnlocked(USER_1_ID) }
+            verify { vaultRepository.isVaultUnlocked(USER_2_ID) }
+            verify { vaultRepository.vaultUnlockDataStateFlow }
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepositoryTest.kt:323]: This is all fine but you can lump these in a single verify block

File: app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt [hunk 9]
```diff
@@ -2,22 +2,37 @@ package com.x8bit.bitwarden.data.platform.processor
...
 /**
  * Default implementation of [BridgeServiceProcessor].
  */
 class BridgeServiceProcessorImpl(
+    private val bridgeRepository: BridgeRepository,
     private val featureFlagManager: FeatureFlagManager,
+    dispatcherManager: DispatcherManager,
 ) : BridgeServiceProcessor {
 
+    private val callbacks by lazy { RemoteCallbackList<IBridgeServiceCallback>() }
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt:33]: Were these necessary just for mocking purposes?

File: app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt [hunk 10]
```diff
@@ -37,42 +52,75 @@ class BridgeServiceProcessorImpl(
      * Default implementation of the bridge service binder.
      */
     private val defaultBinder = object : IBridgeService.Stub() {
-        override fun getVersionNumber(): String {
-            // TODO: BITAU-104
-            return ""
-        }
+
+        override fun getVersionNumber(): String = NATIVE_BRIDGE_SDK_VERSION
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt:56]: 👍

File: app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt [hunk 11]
```diff
@@ -37,42 +52,75 @@ class BridgeServiceProcessorImpl(
...
+            symmetricKeyFingerprint: SymmetricEncryptionKeyFingerprintData?,
         ): Boolean {
-            // TODO: BITAU-104
-            return false
+            if (symmetricKeyFingerprint == null) return false
+            val localSymmetricKeyFingerprint =
+                bridgeRepository.authenticatorSyncSymmetricKey
+                    ?.toSymmetricEncryptionKeyData()
+                    ?.toFingerprint()
+                    ?.getOrNull()
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt:66]: 👍

File: app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt [hunk 12]
```diff
@@ -37,42 +52,75 @@ class BridgeServiceProcessorImpl(
...
+ */
+@Suppress("TooGenericExceptionCaught")
+private fun <T : IInterface> RemoteCallbackList<T>.forEach(action: (T) -> Unit) {
+    val count = this.beginBroadcast()
+    try {
+        for (index in 0..count) {
+            action(this.getBroadcastItem(index))
         }
+    } catch (e: Exception) {
+        // Broadcast failed
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorImpl.kt:122]: @david-livefront I wonder if you think we should add a debug-only error log here. I know that's a direction we're beginning to move. Or a non-fatal firebase log?

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt (hunk 3 above)
Comment (ahaisting-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt:26]: Sweet! c94f35057e01b8fa0ead0a25401afe2721e00eaf

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt [hunk 13]
```diff
@@ -0,0 +1,27 @@
+package com.x8bit.bitwarden.data.platform.repository
+
+import com.bitwarden.bridge.model.SharedAccountData
+
+/**
+ * Provides an API for querying disk sources required by [BridgeRepositoryImpl].
+ *
+ * Note: this repository should not be injected anywhere other than [BridgeRepositoryImpl].
+ */
+interface BridgeRepository {
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt:10]: Are we going to change the naming of all this stuff to `AuthenticatorBridge...` after this PR?

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt [hunk 14]
```diff
@@ -0,0 +1,27 @@
+package com.x8bit.bitwarden.data.platform.repository
+
+import com.bitwarden.bridge.model.SharedAccountData
+
+/**
+ * Provides an API for querying disk sources required by [BridgeRepositoryImpl].
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt:6]: I wouldn't refer to an `...Impl` here. Just say this is required for the Bridge Service functionality?

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepositoryImpl.kt (hunk 4 above)
Comment (ahaisting-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepositoryImpl.kt]: Ah, true. Kinda sad, it means everything inside has to be indented again... but such is life:

eb79ab3e6e833f0de254b6e8fdc2840390426b8e

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+     */
+    val authenticatorSyncSymmetricKey: ByteArray?
+
+    /**
+     * Get a list of shared account data. This function will go through all accounts and for each
+     * one, check to see if the user has Authenticator account syncing enabled and if they
+     * do, it will query and decrypt the user's shared account data.
+     *
+     * Users who do not have authenticator sync enabled or otherwise cannot have their ciphers
+     * accessed will be omitted from the list.
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt:24]: 👍

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+     */
+    suspend fun getSharedAccounts(): SharedAccountData
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt:26]: Do we need to worry about any overall "error" states here? Maybe not?

File: app/src/test/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorTest.kt (hunk 5 above)
Comment (ahaisting-livefront) [reply app/src/test/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorTest.kt:60]: Aye aye!

e9863c266fb2e2f6e132bad665ae34182c8e9649

File: app/src/test/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorTest.kt (hunk 6 above)
Comment (ahaisting-livefront) [reply app/src/test/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorTest.kt:186]: Got u. e9863c266fb2e2f6e132bad665ae34182c8e9649

File: app/src/test/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepositoryTest.kt (hunk 7 above)
Comment (ahaisting-livefront) [reply app/src/test/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepositoryTest.kt]: Indubitably: 94c78b9a54bcb2e83518c9053b5c0137ede1c700

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt (hunk 13 above)
Comment (ahaisting-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt:10]: Yep, absolutely. Here's a ticket for tracking that: https://livefront.atlassian.net/jira/software/c/projects/BITAU/boards/100?selectedIssue=BITAU-172

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt [hunk 17]
```diff
@@ -0,0 +1,27 @@
+package com.x8bit.bitwarden.data.platform.repository
+
+import com.bitwarden.bridge.model.SharedAccountData
+
+/**
+ * Provides an API for querying disk sources required by [BridgeRepositoryImpl].
+ *
+ * Note: this repository should not be injected anywhere other than [BridgeRepositoryImpl].
```
[the remark this answers is not in the document]
Comment (ahaisting-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt:8]: I like that. Updated doc here:

121a762c79d99b01de03b1c78f0f4cebd7e88d96

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt (hunk 14 above)
Comment (ahaisting-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt]: 121a762c79d99b01de03b1c78f0f4cebd7e88d96

File: app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt (hunk 16 above)
Comment (ahaisting-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt:24]: So going through each of the early returns, I don't see any that we are truly expecting that are errors. 

There are a few things that cause an account to omitted from the list, but as we previously discussed, they aren't errors in earnest.

Comment (brian-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/platform/repository/BridgeRepository.kt:24]: OK cool, that sounds good to me then.

File: app/src/test/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorTest.kt [hunk 18]
```diff
@@ -1,52 +1,220 @@
...
+    }
+
+    @Test
+    @Suppress("MaxLineLength")
+    fun `checkSymmetricEncryptionKeyFingerprint should return true if fingerprint does match`() {
+        val binder = getDefaultBinder()
+        every {
+            bridgeRepository.authenticatorSyncSymmetricKey
+        } returns SYMMETRIC_KEY.symmetricEncryptionKey.byteArray
+        assert(binder.checkSymmetricEncryptionKeyFingerprint(SYMMETRIC_KEY_FINGERPRINT))
```
Comment (david-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorTest.kt:120]: One more!

Comment (ahaisting-livefront) [reply app/src/test/java/com/x8bit/bitwarden/data/platform/processor/BridgeServiceProcessorTest.kt]: Thanks!

8b7f522455e05bdf19503834fc5c629834737c90

**Review by david-livefront — APPROVED** [review]

**Comment by ahaisting-livefront** [comment]
Thanks, @brian-livefront and @david-livefront !

**Pull request merged**